src/tools/compiletest/src/runtest.rs

```diff
@@ -2375,9 +2375,13 @@ impl<'test> TestCx<'test> {
         let rust_src_dir = rust_src_dir.read_link_utf8().unwrap_or(rust_src_dir.to_path_buf());
         normalize_path(&rust_src_dir.join("library"), "$SRC_DIR_REAL");
 
+        // Canonicalize test build directory path.
+        // Without this some tests fail if build directory is a symlink.
+        let output_base_dir = fs::canonicalize(self.output_base_dir()).unwrap();
```

Comment:
Suggestion: I just switched compiletest over to use `camino` to enforce UTF-8 paths, so this should instead be

```rs
let output_base_dir = self.output_base_dir().canonicalize_utf8().unwrap();
```

Reply:
Discussion: that being said, I'm not 100% sure this is sufficient in itself, since rustc itself might have inconsistencies when `build/` is a symlink to elsewhere...

Reply:
(note that there are more `fs::canonicalize` in the file, not sure if they are intended)

The tests seem to pass for me, so this is *probably* fine. I'm more concerned about a case where the directory doesn't exist, or, windows where canonicalize might return weird paths:
> On Windows, this converts the path to use [extended length path](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file) syntax, which allows your program to use longer path names, but means you can only join backslash-delimited paths to it, and it may be incompatible with other applications (if passed to the application on the command-line, or written to a file another application may read).

Reply:
Yeah, some of these runtest bits need further auditing, which is beyond the scope of this PR.